### DISCUSSION TITLE
Does Freeform always set Cache-Control: no-store, no-cache, must-revalidate?

### DISCUSSION THREAD

**Discussion by terryupton** [solspace/craft-freeform#973] [Questions & Answers] [answered]
**Describe the bug or issue you're experiencing**
It seems that any page in Craft that contains a freeform form.
It sets a header of `Cache-Control: no-store, no-cache, must-revalidate`

I have managed to pin this down after checking several urls and testing sites.

Is this correct and why? Is there a way to bypass this...for using Cloudflare or CGI caching?

###Intentions

So I am attempting to test and setup some Cloudflare on Cladwinds (dev subdomain) (dev.cladwinds.com).
So in order to test and play with Cloudflare cache working along side Freeform. I have done this on the sub-domain initially. Caching everything in Cloudflare.
However, before I get into trying to test the pages with a form and then find the solution for this, I wanted to check that the caching was working!? 

However,  it isn't working. 
Showing as BYPASS in the CF-cache headers. 
<img width="619" alt="Screenshot 2024-01-07 at 18 41 13" src="https://github.com/solspace/craft-freeform/assets/5698596/68d71b23-ae2b-4347-b7b8-0aab4cc51c84">

My setup is the same as I have done for another Craft site that has no freeform and this works fine.

So I looked further and I am seeing that the Origin server is setting:
Cache-Control: no-store, no-cache, must-revalidate

Obviously this was not being set on the davidnewens.com website.
All my Craft websites are pretty similar in their build / framework and plugins. 

I have a feeling it is something Craft is also setting this when logged into the CMS.
But I think Freeform is setting this indefinitely on any page that contains a Freeform form.

**Comment by kjmartens** [comment]
I checked with a developer and he says that we're not setting these headers at all. We only pass them when making a form POST via AJAX. We're also setting cache headers for the Freeform plugin CSS and JS, but those are set to use cache, not skip cache. 🙂

**Comment by timkelty** [comment]
> So I looked further and I am seeing that the Origin server is setting:
> Cache-Control: no-store, no-cache, must-revalidate

![giphy](https://github.com/solspace/craft-freeform/assets/18329/d216303e-58fc-4038-8536-08eea3ffaadf)

Those headers likely indicate a `session_start` is being called. PHP will set headers based on https://www.php.net/manual/en/session.configuration.php#ini.session.cache-limiter

Related discussion: https://craftcms.stackexchange.com/questions/2598/why-does-craft-set-no-cache-headers-for-all-requests

**Comment by terryupton** [comment]
Thanks @timkelty and @kjmartens 
So what's the solution here? 

I want to try and cache the pages and ultimately try and implement this as per the docs. https://docs.solspace.com/craft/freeform/v4/templates/caching/

But it would seem this isn't possible? So how are others doing this, if you cannot cache pages that contain a session, which I assume is added because of the Freeform. 

I don't see these headers at all on pages that don't contain a Freeform. So it's not being set explicitly by the origin server or craft. 

Any help and advice appreciated. 

Thanks

**Comment by seandelaney** [comment]
We are not setting any cache skip related headers in Freeform. However, I still had a look at the Freeform codebase anyways (as I got paranoid) and I can’t find anything.

If I view a CraftCMS site with Freeform installed though, I can see `no-store, no-cache, must-revalidate` set!!

So I uninstalled Freeform from CraftCMS and I still see `no-store, no-cache, must-revalidate` being set, so something is setting these but it is not coming from Freeform as I’ve completely removed the plugin - uninstalled and then deleted the files. In fact, I completely removed all plugins I had installed in this test CraftCMS site.

If I search the CraftCMS codebase on Github, I can see where the CraftCMS sets cache control headers - its when you are logged into the CP. I checked the CP response and I can see `no-store, no-cache, must-revalidate`. 

But I can confirm Freeform is not setting these. But what is?

Then I thought, maybe if you were logged into the CP but also viewed the frontend (either in same tab or different tab), that the authenticated session/cookie/response info was carried over? So I logged out of the CP and viewed the frontend again. I still saw `no-store, no-cache, must-revalidate`. I tried a different browser - Firefox, where I’ve never logged into the CP before. Viewed the frontend and got no cache control headers set. Bingo. I went back to my original browser, Brave and I cleared cookies, session + local storage, refreshed the frontend again, and now I see no cache control headers being set. So there we have it. If you are logged into the CP and viewing the frontend, `no-store, no-cache, must-revalidate` are being applied to the response, because it knows you are also logged into the CP.

You can try it yourself.

Final test. I then installed Freeform again. Viewed frontend demo templates while logged into CP. Back to seeing `no-store, no-cache, ...`. 

![image](https://github.com/solspace/craft-freeform/assets/580403/5330d835-33bb-41ed-a688-8d4758ca0d3c)

I logged out, clear cookies, session info etc from browser. Viewed the frontend demo templates again. No cache control headers set:

![image](https://github.com/solspace/craft-freeform/assets/580403/5ed685b6-2f81-431f-95b8-bc1d183dc0b2)

**Comment by kjmartens** [comment, marked as answer]
@terryupton,

Sorry about all this... it looks like there was something we could change in Freeform to help with this:
https://github.com/solspace/craft-freeform/pull/1144 (thank you @timkelty!)

We'll have some updates out shortly. 🙂
